core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/FileAsyncRequestBodySplitHelper.java

```diff
@@ -0,0 +1,162 @@
+
+    private volatile boolean isDone = false;
+
+    private AtomicInteger numAsyncRequestBodiesInFlight = new AtomicInteger(0);
+    private AtomicInteger chunkIndex = new AtomicInteger(0);
+
+    public FileAsyncRequestBodySplitHelper(FileAsyncRequestBody asyncRequestBody,
+                                           AsyncRequestBodySplitConfiguration splitConfiguration) {
+        Validate.notNull(asyncRequestBody, "asyncRequestBody");
+        Validate.notNull(splitConfiguration, "splitConfiguration");
+        Validate.isTrue(asyncRequestBody.contentLength().isPresent(), "Content length must be present", asyncRequestBody);
+        this.totalContentLength = asyncRequestBody.contentLength().get();
+        this.remainingBytes = new AtomicLong(totalContentLength);
+        this.path = asyncRequestBody.path();
+        this.chunkSize = splitConfiguration.chunkSizeInBytes() == null ?
+                         AsyncRequestBodySplitConfiguration.defaultConfiguration().chunkSizeInBytes() :
+                         splitConfiguration.chunkSizeInBytes();
+        this.totalBufferSize = splitConfiguration.bufferSizeInBytes() == null ?
+                               AsyncRequestBodySplitConfiguration.defaultConfiguration().bufferSizeInBytes() :
+                               splitConfiguration.bufferSizeInBytes();
+        this.bufferPerAsyncRequestBody = asyncRequestBody.chunkSizeInBytes();
+    }
+
+    public SdkPublisher<AsyncRequestBody> split() {
+
+        SimplePublisher<AsyncRequestBody> simplePublisher = new SimplePublisher<>();
+
+        try {
+            sendAsyncRequestBody(simplePublisher);
+        } catch (Throwable throwable) {
+            simplePublisher.error(throwable);
+        }
+
+        return SdkPublisher.adapt(simplePublisher);
+    }
+
+    private void sendAsyncRequestBody(SimplePublisher<AsyncRequestBody> simplePublisher) {
+        if (!isSendingRequestBody.compareAndSet(false, true)) {
+            return;
+        }
+
+        try {
+            doSendAsyncRequestBody(simplePublisher);
+        } finally {
+            isSendingRequestBody.set(false);
+        }
+    }
+
+    private void doSendAsyncRequestBody(SimplePublisher<AsyncRequestBody> simplePublisher) {
+        while (true) {
+            if (!shouldSendMore()) {
+                break;
+            }
+
+            AsyncRequestBody currentAsyncRequestBody = newFileAsyncRequestBody(simplePublisher);
+            simplePublisher.send(currentAsyncRequestBody);
+            numAsyncRequestBodiesInFlight.incrementAndGet();
+            checkCompletion(simplePublisher, currentAsyncRequestBody);
+        }
+    }
+
+    private void checkCompletion(SimplePublisher<AsyncRequestBody> simplePublisher, AsyncRequestBody currentAsyncRequestBody) {
+        long remaining = remainingBytes.addAndGet(-currentAsyncRequestBody.contentLength().get());
+
+        if (remaining == 0) {
+            isDone = true;
+            simplePublisher.complete();
+        }
```

Comment:
Can we add a branch for < 0 and call error on the publisher? If there's a bug, I'd rather it manifest as an exception than a hung future.